pkg/controllers/cloudmap_controller.go

```diff
@@ -208,37 +208,114 @@ func (r *CloudMapReconciler) createAndGetDerivedService(ctx context.Context, svc
 	return r.getDerivedService(ctx, svc.Namespace, svcImport.Annotations[DerivedServiceAnnotation])
 }
 
-func (r *CloudMapReconciler) updateEndpointSlices(ctx context.Context, svcImport *v1alpha1.ServiceImport, cloudMapService *model.Service, svc *v1.Service) error {
+func (r *CloudMapReconciler) updateEndpointSlices(ctx context.Context, svcImport *v1alpha1.ServiceImport, desiredEndpoints []*model.Endpoint, svc *v1.Service) error {
 	existingSlicesList := discovery.EndpointSliceList{}
-	var existingSlices []*discovery.EndpointSlice
 	if err := r.Client.List(ctx, &existingSlicesList,
 		client.InNamespace(svc.Namespace), client.MatchingLabels{discovery.LabelServiceName: svc.Name}); err != nil {
 		return err
 	}
-	if len(existingSlicesList.Items) == 0 {
-		// create new endpoint slice
-		existingSlices = createEndpointSlicesStruct(svcImport, cloudMapService, svc)
-		for _, slice := range existingSlices {
-			if err := r.Client.Create(ctx, slice); err != nil {
+
+	desiredPorts := extractEndpointPorts(desiredEndpoints)
+	matchedEndpoints := make(map[string]*discovery.Endpoint)
+	endpointsToCreate := make([]discovery.Endpoint, 0)
+
+	// check if all desired endpoints are in an endpoint slice already
+	for _, desiredEndpoint := range desiredEndpoints {
+		existingEndpoint := findEndpointInSliceByIp(existingSlicesList.Items, desiredEndpoint.IP)
+		if existingEndpoint != nil {
+			matchedEndpoints[desiredEndpoint.IP] = existingEndpoint
+		} else {
+			endpointsToCreate = append(endpointsToCreate, createEndpointForSlice(svc, desiredEndpoint.IP))
+		}
+	}
+
+	// check if all endpoints in slices match a desired endpoint,
+	for _, existingSlice := range existingSlicesList.Items {
+		updatedEndpointList := make([]discovery.Endpoint, 0)
+		for _, existingEndpoint := range existingSlice.Endpoints {
+			keep, found := matchedEndpoints[existingEndpoint.Addresses[0]]
+			if found {
+				updatedEndpointList = append(updatedEndpointList, *keep)
+			}
+		}
+
+		dirty := len(existingSlice.Endpoints) != len(updatedEndpointList)
+
+		// fill endpoint slice with endpoints to create if necessary and there is sufficient room
+		for _, endpointToCreate := range endpointsToCreate {
+			if len(updatedEndpointList) >= 100 {
```

Comment:
We can have upto [1000](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#endpointslice-resource) endpoints per endpoint slice by configuring `--max-endpoints-per-slice` flag. Are we planning to support this in the future.
Maybe, we can also add comment about why this magic number is used. And extract it out as a constant.

Reply:
I don't know if slices that large would be preferable over the previous endpoints api then, or if we'd gain anything by doing that. I think it's something we can explore during load test, make it optionally configurable. I just took the default recommended slice size from the docs, so I don't have an explanation to add, but I agree it should be a constant at least.